### PR TITLE
[IMP] pricelist: pricelist revamp

### PR DESCRIPTION
**Pull request by livaios**
- no more discount policy -> only show discount if compute price of rule is percentage
- show discounted value automatically on website
- no more "multiple prices per product" advanced settings turned on by default
- move pricelist settings related to website to website_sale and delete configuration tab

task-3061255

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/data/demo.xml [hunk 1]
```diff
@@ -271,8 +271,7 @@
 
         <record id="list_christmas" model="product.pricelist">
             <field name="name">Christmas</field>
-            <field name="selectable" eval="False" />
-            <field name="website_id" ref="website.default_website" />
+            <field name="website_ids" eval="[(4, ref('website.default_website'))]"/>
```
Comment (Feyensv) [inline addons/website_sale/data/demo.xml:274]: FYI, you can use Command.link in xml files too (it's available in the evaluation environment)

File: addons/website_sale/models/product_pricelist.py [hunk 2]
```diff
@@ -9,44 +9,29 @@
...
-    @api.constrains('company_id', 'website_id')
+    @api.constrains('company_id', 'website_ids')
     def _check_websites_in_company(self):
         """ Prevent misconfiguration multi-website/multi-companies.
 
         If the record has a company, the website should be from that company.
         """
-        for record in self.filtered(lambda pl: pl.website_id and pl.company_id):
-            if record.website_id.company_id != record.company_id:
+        for record in self.filtered(lambda pl: pl.website_ids and pl.company_id):
```
Comment (Feyensv) [inline addons/website_sale/models/product_pricelist.py:33]:
```suggestion
        for pricelist in self.filtered(lambda pl: pl.website_ids and pl.company_id):
```

File: addons/website_sale/models/product_pricelist.py [hunk 3]
```diff
@@ -122,10 +106,18 @@ def _get_website_pricelists_domain(self, website):
...
             ('active', '=', True),
             ('company_id', 'in', [False, website.company_id.id]),
-            '|', ('website_id', '=', website.id),
-            '&', ('website_id', '=', False),
-            '|', ('selectable', '=', True), ('code', '!=', False),
+            '|', '|', ('website_ids', 'any', [('id', '=', website.id)]),
+            '&', ('website_ids', '=', False),
+            ('code', '!=', False),
+            # If default pricelist
+            ('name', '=', default_pricelist['name']),
```
Comment (Feyensv) [inline addons/website_sale/models/product_pricelist.py:118]: unreliable as it'll change according to the current user/website lang, why do you need the default pricelist ?

File: addons/website_sale/models/product_template.py [hunk 4]
```diff
@@ -548,7 +544,11 @@ def _get_additionnal_combination_info(self, product_or_template, quantity, date,
...
 
-        if pricelist.discount_policy != 'without_discount':
+        if pricelist._get_product_rule_policy(
+            product=product_or_template,
+            quantity=quantity,
+            target_currency=currency,
+        ) != 'percentage':
             # Leftover from before cleanup, different behavior between ecommerce & backend configurator
             # probably to keep product sales price hidden from customers ?
             combination_info['list_price'] = combination_info['price']
```
Comment (Feyensv) [inline addons/website_sale/models/product_template.py:554]: This part could be dropped if discount computation is harmonized between backend and fronted (cf other comment).

File: addons/website_sale/models/res_config_settings.py [hunk 5]
```diff
@@ -92,6 +92,13 @@ class ResConfigSettings(models.TransientModel):
     enabled_extra_checkout_step = fields.Boolean(string="Extra Step During Checkout")
     enabled_buy_now_button = fields.Boolean(string="Buy Now")
 
+    selectable_pricelist_ids = fields.Many2many(
+        comodel_name='product.pricelist',
+        relation='selectable_pricelists',
```
Comment (Feyensv) [inline addons/website_sale/models/res_config_settings.py:97]:
```suggestion
```

Is the relation necessary? it won't be stored on the wizard model anyway :thinking: 
Also, the comodel_name is automatically deducted from the `related` field.

File: addons/website_sale/models/res_config_settings.py [hunk 6]
```diff
@@ -110,6 +117,24 @@ def _inverse_account_on_checkout(self):
...
+            selected_pricelists = ProductPricelist.search([
+                ('id', 'in', res_config.selectable_pricelist_ids.ids),
+            ])
+            for pricelist in selected_pricelists:
+                pricelist.website_ids |= website_id
+            unselected_pricelists = ProductPricelist.search(
+                [('website_ids', 'any', [('id', '=', website_id.id)])]
+            ) - selected_pricelists
+            for pricelist in unselected_pricelists:
+                pricelist.website_ids -= website_id
```
Comment (Feyensv) [inline addons/website_sale/models/res_config_settings.py:136]: Never change another model in an onchange !
It means that even if you don't save your modified settings, it'd still have had an impact on existing records already.

File: addons/website_sale/models/website.py [hunk 7]
```diff
@@ -160,7 +160,13 @@ def _default_recovery_mail_template(self):
     all_pricelist_ids = fields.One2many(
         string="All pricelists",
         comodel_name='product.pricelist',
-        inverse_name='website_id',
+        inverse_name='website_ids',
+    )
```
Comment (Feyensv) [inline addons/website_sale/models/website.py:164]: Won't work, O2M should be paired with a M2O, you should remove this field.

File: addons/website_sale/models/website.py [hunk 8]
```diff
@@ -226,11 +232,12 @@ def _get_pl_partner_order(
...
-            # Only show selectable or currently used pricelist (cart or session)
-            check_pricelist = lambda pl: pl.selectable or pl.id in (current_pl_id, order_pl_id)
-        else:
-            check_pricelist = lambda _pl: True
+        # Only show selectable or currently used pricelist (cart or session)
+        def check_pricelist(pl):
+            if show_visible:
+                return pl._is_selectable() or pl.id in (current_pl_id, order_pl_id)
+            else:
+                return True
```
Comment (Feyensv) [inline addons/website_sale/models/website.py:240]: Why did you change the way the method is created here? :thinking:

**Review by Feyensv — COMMENTED** [review]
Took the liberty to take an early look as it's a big task :eyes:
Didn't go too deep in the changes, but still noticed some things where I had remarks/questions (since pricelists are a model I worked a lot on :D).

Also, please keep your PR in draft until it has at least been through Testing IN and Peer review, to avoid pinging people on unfinished work.

File: addons/product/models/product_pricelist.py [hunk 9]
```diff
@@ -160,6 +150,27 @@ def _get_product_rule(self, product, *args, **kwargs):
         self and self.ensure_one()  # self is at most one record
         return self._compute_price_rule(product, *args, compute_price=False, **kwargs)[product.id][1]
 
+    def _get_product_rule_policy(self, product, *args, **kwargs):
+        """Compute the pricelist price & rule for the specified product, qty & uom.
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_pricelist.py:154]: Only on sales order line for now, we'll have to find another solution for configurator & ecommerce logic.

File: addons/point_of_sale/models/pos_order.py [hunk 10]
```diff
@@ -1169,6 +1169,7 @@ class PosOrderLine(models.Model):
         ('manual', 'Manual'),
         ('automatic', 'Automatic'),
     ], string='Price Type', default='original')
+    pricelist_item_id = fields.Many2one('product.pricelist.item')
```
Comment (Feyensv) [inline addons/point_of_sale/models/pos_order.py:1172]: where do you set that field?

File: addons/product/models/product_pricelist_item.py [hunk 11]
```diff
@@ -112,10 +124,16 @@ def _default_pricelist_id(self):
              "Rounding is applied after the discount and before the surcharge.\n"
              "To have prices that end in 9.99, set rounding 10, surcharge -0.01")
     price_surcharge = fields.Float(
-        string="Price Surcharge",
+        string="Extra Fee",
         digits='Product Price',
         help="Specify the fixed amount to add or subtract (if negative) to the amount calculated with the discount.")
 
+    price_discount_surcharge = fields.Float(
```
Comment (Feyensv) [inline addons/product/models/product_pricelist_item.py:131]: does it happen to use both `price_surcharge` and `price_discount_surcharge` or could we use one field for the two logic ?

File: addons/product/models/product_pricelist_item.py [hunk 12]
```diff
@@ -438,8 +493,8 @@ def _compute_base_price(self, product, quantity, uom, date, currency):
         return price
 
     def _compute_price_before_discount(self, *args, **kwargs):
-        """Compute the base price of the lowest pricelist rule whose pricelist discount_policy
-        is set to show the discount to the customer.
+        """Compute the base price of the lowest pricelist rule,
+        discount is shown by default if computation method is a percentage rule.
```
Comment (Feyensv) [inline addons/product/models/product_pricelist_item.py:497]: Depending on dala's answer, we might want to drop this method & logic (only keep the last discount amount)

**Review by Feyensv — COMMENTED** [review]
even more partial review comments :eyes:

(didn't have the time to discuss about this task with you and dala last week, maybe this week :D)

**Comment by chapital** [comment]
Please also consider this.

#149330

File: addons/product/models/product_pricelist_item.py [hunk 13]
```diff
@@ -221,6 +263,13 @@ def _check_product_consistency(self):
                 raise ValidationError(_("Please specify the product variant for which this rule should be applied"))
 
     #=== ONCHANGE METHODS ===#
+    @api.onchange('base')
```
Comment (Feyensv) [inline addons/product/models/product_pricelist_item.py:266]:
```suggestion

    @api.onchange('base')
```

File: addons/product/models/product_pricelist_item.py [hunk 14]
```diff
@@ -186,6 +220,14 @@ def _compute_rule_tip(self):
...
             )
 
+    def _get_integer(self, percentage):
+        return int(percentage) if percentage.is_integer() else percentage
+
+    def _get_displayed_discount(self, item):
+        if item.base == 'standard_price':
+            return "markup", self._get_integer(item.price_markup)
+        return "discount", self._get_integer(item.price_discount)
+
```
Comment (Feyensv) [inline addons/product/models/product_pricelist_item.py:230]: not translated ? :thinking:

Comment (Feyensv) [reply addons/product/models/product_pricelist_item.py:230]: There should be no pot update in master, it'll be generated after the freeze for all the modules.
I meant that you're not using the translation method here.

Comment (Feyensv) [reply addons/product/models/product_pricelist_item.py]: Yes, or it won't be translated ;)

**Comment by Feyensv** [comment]
@robodoo r+ rebase-ff

**Comment by ploegvde** [comment]
@livaios Ok, but If you now have this scenario in odoo 17:
Product price = 100
Discount in price list = 10%
Unit price = 90 on sale order line / discount percentage is 0 (because discount policy is 'Discount included in the price'
There are even Odoo users that have removed the discount column from the report with Studio

In Odoo 18 it will be this (Am I correct?)
Product price = 100
Discount in price list = 10%
Unit price = 100 on sale order line / discount percentage is 10%

So, with this change the receiving customer of a quote gets a different overview of the price (Yes nett it is same).

**Comment by ploegvde** [comment]
@livaios Ok, but how do we manage that if customers are using that? Why force customers to change the pricing policy?

**Comment by dalaOdoo** [comment]
Hi @ploegvde can you explain why you consider showing the pricelist discount to the customer is a problem ? 

The work around will be to define the unit price in the customer pricelist instead of using a discount.

**Comment by ploegvde** [comment]
@dalonsod I know cases where the standard price for a market segment is too high. But showing the discount makes it visible that they are basically expensive. In this case, you make a price list with a discount, without showing the discount to the ens customer. 

If you serve multiple market segments, and you want to give them all a separate price based on a base price (so a discount on the standard product price) you have to do it like this. You would rather not show every segment a discount.

It is not always good to show discounts. Discounts can give the wrong impression of the company. But using (internally) discounts makes life easier. A change in base price also changes the specific price list price.

Why would you drop the possibility? And again, how to manage this in migrations to Odoo 18?

**Comment by dalonsod** [comment]
@ploegvde you might call for @dalaOdoo 😅

**Comment by ploegvde** [comment]
@dalonsod  Oops, sorry

**Comment by chapital** [comment]
Hi, what happens if the discount is negative?

For example I use the discount field to add the revenue, for example:

Unit price: 100
Discount: -10%
Net total: 110

**Comment by dalaOdoo** [comment]
@chapital it's not shown as a discount.

**Comment by dalaOdoo** [comment]
@ploegvde 

We are going to make some small changes:
- The discount will be shown to customers only if compute_price == percentage (aka Discount) and if the discount feature on SOL is active. 
- It will be possible to select a price list as the base price for a discount rules.

It means if a user don't want to show the discount to a customer he can use formula to compute the price (or not activate the discount feature on SOL)

**Comment by ploegvde** [comment]
@dalaOdoo Thanks, think that is a suitable solution. I'm only curious how Odoo handles these changes with existing customers that migrate to Odoo 18 and are confronted (if they see it) with new output on the quote?

**Comment by ploegvde** [comment]
@dalaOdoo Truly interested in your response :)

**Comment by Feyensv** [comment]
@ploegvde ATM, we only plan to document the change in the release notes.  No specific conversion of existing data/configuration has been planned.
Sometimes, changes are made for what we believe to be a better future, even if some customers may be temporarily inconvenienced by the changes while they adapt to it.

FYI, the pr with the small changes that Damien spoke about is #173149.

Have a nice day.

**Comment by ploegvde** [comment]
@Feyensv Thanks for the kind reply. I can only hope for clear relaese notes, so that this changes doesn't impact users a lot. If it was me, I would add it to the migration report, so people can act on it.

**Comment by RFoerster** [comment]
Hi @dalaOdoo and @Feyensv ,
could you please confirm the following usecase:

We have a client that has list-prices in their product (i.e. 100$ net). The price is then discounted with a pricelist but it is not discounted in percent but to a fixed price (i.e. 90$). With v17 it was possible to show the 100$ as strike-through price and sell the product for 90$ in PoS. Now with the new functionality, having fixed-prices and showing the strike-through-price is not possible anymore in PoS.

Could you please confirm that? Is there any workaround?
Also please notice there is almost no information about that in the release notes. It only says "Pricelist views and usages have been simplified".

Thank you!

-Robert
